### ISSUE TITLE
Generic Multiset Programming as a Relational Engine?

### ISSUE THREAD

**Issue by gabriel-fallen** [agentm/project-m36#17]
Hi!

I'm sorry, this isn't really an issue rather discussion question, but I couldn't find mail list or something like that.

Have you considered this approach - https://www.researchgate.net/publication/220607041_Generic_multiset_programming_with_discrimination-based_joins_and_symbolic_Cartesian_products - as a relational engine?
It seams reasonably efficient and better integrated into functional programming.

I'm sorry if I ask stupid questions... :)

**Comment by agentm** [comment]
@gabriel-fallen Thanks for linking to this interesting paper!

The paper first demonstrates how to use ADTs to perform static optimization. We have implemented some basic static optimizations in the [same fashion](https://github.com/agentm/project-m36/blob/master/ProjectM36/StaticOptimizer.hs). Indeed, there are many similarities in the algebraic data types we have independently implemented, though that is not particularly surprising. 

Clearly, the paper focuses on implementing SQL constructs, thus, it implements optimizations for multisets (bags) of tuples rather than relations (set of tuples). In several places, the paper actually points out how the optimizations would be easier if they did not have to take multisets into account (e.g. p.17), which is consistent with our own findings. One thing the paper does do differently is execute its optimizations at evaluation time:

> Note that optimization occurs during evaluation; there is no
separate query optimization phase. The cost of doing this on the fly
is limited even for moderately sized input data—only very simple
rules, all related to processing of formal products and unions, are
executed. In contrast to techniques for static query optimization
(optimizations performed on queries independent of the actual data
the query is applied to) doing it at run-time provides the opportunity
for performing data dependent query optimization. (This is not
illustrated here, however.)

I think that this is an important point and perhaps that Project:M36 does incorrectly now, though the end result is currently the same. It is definitely valuable to allow optimizations to be deferred as well, especially when considering distributed databases where one database may not have full or up-to-date insight into another database's execution costs.

From my own perspective, the most valuable part of the paper is the reference to [a paper describing how to execute joins in linear memory space](http://www.diku.dk/hjemmesider/ansatte/henglein/papers/henglein2010.pdf) by the same author. I found (a video)[https://vimeo.com/16754831] of the author summarizing his paper as well. The second paper introduces "generic equivalence discrimination" which elucidates on this strategy.

The paper appears to rediscover [dlists](https://hackage.haskell.org/package/dlist) in a set context, which is less interesting.

There is much overlap in the paper's design with the Project:M36 design, which increases our confidence in our own design. 

The following paragraph sums up what could be most useful to Project:M36:

> Discrimination seems to be unknown or at least unused in
databases. The core techniques for ‘bringing items that are “alike”
together and performing some operation on them’ (Graefe 1993,
Section 2) in database systems are sorting and hashing. When re-
stricted to base types such as strings, discrimination mostly resem-
bles radix sorting, but performs only partitioning. In that sense it is
reminiscent of hashing, but it provides complete (exact) partitioning
in terms of the join classification of Mishra and Eich (1992, Section
4), where hashing provides only disjoint partitioning, the quality of
which depends on the actual data distribution. What this means is
that after the partitioning phase the match step in a discrimination-
based join is trivial: it consists of the union of the cross-products
of corresponding partition blocks. In a hashing-based join corre-
sponding blocks have to be postprocessed since not all pairs from
the cross-product of corresponding partition blocks are necessarily
in the result.

In any case, I will contact the author for potential discussion/collaboration. Thanks for the link!
